### PR TITLE
Allow RACCommand to specify generic input type

### PR DESCRIPTION
**Pull request by erichoracek**
Inputs to `RACCommand` to do not suffer from the same issues as `RACSignal` when attempting to adopt lightweight generics as discussed in #2085. As such, adding them is as simple as declaring a `<__covariant InputType>` in the interface.

In projects that are still using RAC 2.X, this will be helpful to ensure that commands are executed with an input of the correct type. It also helps to catch programmer error when refactoring commands.

I was unable to find a way to expose this annotated type to the compiler when bridging a `RACCommand` to `Action` in `ObjectiveCBridging.swift`. Open to any suggestions on how this might be possible.

### CODE REVIEW

**Comment by NachoSoto** [comment]
This is great, thank you!

As for bridging this type back to Swift, I don't think it's possible right now to inherit the generic type for `Action`.

**Comment by robb** [comment]
LGTM

**Comment by erichoracek** [comment]
Seems that if the [Importing Objective-C Lightweight Generics](https://github.com/apple/swift-evolution/blob/master/proposals/0057-importing-objc-generics.md) proposal is accepted we may be able to supply the generic input to the bridged `Action` in `ObjectiveCBridging.swift`!

Regardless, is there anything preventing this from being merged as-is? Thanks!

File: ReactiveCocoa/Objective-C/RACCommand.h [hunk 1]
```diff
@@ -24,7 +24,7 @@ extern NSString * const RACUnderlyingCommandErrorKey;
 
 /// A command is a signal triggered in response to some action, typically
 /// UI-related.
-@interface RACCommand : NSObject
+@interface RACCommand<__covariant InputType> : NSObject
```
Comment (mdiep) [inline ReactiveCocoa/Objective-C/RACCommand.h]: Shouldn't `InputType` be contravariant?

You've specified that `NSControl.rac_command` should be `RACCommand<__kindof NSControl *>`, but I should be able to assign a `RACCommand<NSObject *>` to that property. That command needs to be able to take an `NSControl`, but it needn't accept only `NSControl`s. Conversely, it wouldn't be right to assign a `RACCommand<NSButton *>` to it, because `NSControl` can't guarantee that an `NSButton` will be sent.

In other words, `RACCommand` inputs behave like function parameters, which are contravariant.

Comment (erichoracek) [inline ReactiveCocoa/Objective-C/RACCommand.h]: @mdiep Great point! I was unaware that multiple variance keywords were even available. After reading up on this the suggested change definitely makes sense. Strange that this is available in Obj-C but not in Swift. Thanks!

File: ReactiveCocoa/Objective-C/NSControl+RACCommandSupport.h [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 #import <Cocoa/Cocoa.h>
 
-@class RACCommand;
+#import "RACCommand.h"
```
Comment (ikesyo) [inline ReactiveCocoa/Objective-C/NSControl+RACCommandSupport.h]: It seems that we can write as `@class RACCommand<__contravariant InputType>;` here.

**Comment by mdiep** [comment]
Thanks! And sorry for the delay.

In the future, you should make sure that you leave a comment after you push so that I get a notification. I think that's why I missed this.

**Comment by ikesyo** [comment]
Should this be backported to `2.x-maintenance`?

**Comment by ikesyo** [comment]
> Should this be backported to `2.x-maintenance`?

Done in b4a166f.
